se_denseslam/include/se/rendering.hpp

```diff
@@ -0,0 +1,202 @@
+ CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __RENDERING_HPP
+#define __RENDERING_HPP
+
+#include <tuple>
+#include <sophus/se3.hpp>
+
+#include <timings.h>
+#include <se/utils/math_utils.h>
+#include <se/commons.h>
+#include <se/continuous/volume_template.hpp>
+#include <se/image/image.hpp>
+#include <se/ray_iterator.hpp>
+
+/* Raycasting implementations */
+#include <se/bfusion/rendering_impl.hpp>
+#include <se/kfusion/rendering_impl.hpp>
+
+template <typename T>
+using Volume = VolumeTemplate<T, se::Octree>;
+
+template<typename T>
+void raycastKernel(const Volume<T>&            volume,
+                   se::Image<Eigen::Vector3f>& vertex,
+                   se::Image<Eigen::Vector3f>& normal,
+                   const Eigen::Matrix4f&      view,
+                   const float                 nearPlane,
+                   const float                 farPlane,
+                   const float                 mu,
+                   const float                 step,
+                   const float                 largestep) {
+  TICK();
+#pragma omp parallel for
+  for (int y = 0; y < vertex.height(); y++)
+#pragma simd
+    for (int x = 0; x < vertex.width(); x++) {
+
+      Eigen::Vector2i pos(x, y);
+      const Eigen::Vector3f dir =
+        (view.topLeftCorner<3, 3>() * Eigen::Vector3f(x, y, 1.f)).normalized();
+      const Eigen::Vector3f transl = view.topRightCorner<3, 1>();
+      se::ray_iterator<T> ray(*volume._map_index, transl, dir, nearPlane, farPlane);
+      ray.next();
+      const float t_min = ray.tcmin(); /* Get distance to the first intersected block */
+      const Eigen::Vector4f hit = t_min > 0.f ?
+        raycast(volume, transl, dir, t_min, ray.tmax(), mu, step, largestep) :
+        Eigen::Vector4f::Constant(0.f);
+      if (hit.w() > 0.0) {
+        vertex[x + y * vertex.width()] = hit.head<3>();
+        Eigen::Vector3f surfNorm = volume.grad(hit.head<3>(),
+            [](const auto& val){ return val.x; });
+        if (surfNorm.norm() == 0) {
+          //normal[pos] = normalize(surfNorm); // APN added
+          normal[pos.x() + pos.y() * normal.width()] = Eigen::Vector3f(INVALID, 0, 0);
+        } else {
+          // Invert normals if SDF
+          normal[pos.x() + pos.y() * normal.width()] = std::is_same<T, SDF>::value ?
+            (-1.f * surfNorm).normalized() : surfNorm.normalized();
+        }
+      } else {
+        vertex[pos.x() + pos.y() * vertex.width()] = Eigen::Vector3f::Constant(0);
+        normal[pos.x() + pos.y() * normal.width()] = Eigen::Vector3f(INVALID, 0, 0);
+      }
+    }
+  TOCK("raycastKernel", vertex.width() * vertex.height());
+}
+
+// void renderNormalKernel(uchar3* out, const float3* normal, uint2 normalSize) {
```

Comment:
Since we are here, why don't you remove commented code? 
Thanks!

Reply:
Removed commented out code in rendering.hpp and tracking.cpp